### ISSUE TITLE
Wrong batch size reported when using TPUStrategy

### ISSUE THREAD

**Issue by bjourne** [tensorflow/tensorflow#40428]
In the summary, batch size is reported as 4 but I specified it as 32:
```python
from os import environ
from tensorflow.config import experimental_connect_to_cluster
from tensorflow.distribute.cluster_resolver import TPUClusterResolver
from tensorflow.distribute.experimental import TPUStrategy
from tensorflow.keras import Model
from tensorflow.keras.layers import *
from tensorflow.tpu.experimental import initialize_tpu_system
import tensorflow as tf

tpu_addr = environ.get('COLAB_TPU_ADDR')
resolver = TPUClusterResolver('grpc://' + tpu_addr)
experimental_connect_to_cluster(resolver)
initialize_tpu_system(resolver)
strategy = TPUStrategy(resolver)

with strategy.scope():
    source = Input(shape = (320,),
                   batch_size = 32,
                   dtype = tf.int32)
    embedding = Embedding(input_dim = 49, output_dim = 100)(source)
    lstm_1 = LSTM(700,
                  return_sequences = True,
                  dropout = 0.3)(embedding)
    y_pred = TimeDistributed(
        Dense(49, activation = 'softmax'))(lstm_1)
    m = Model(inputs = [source], outputs = [y_pred])
    m.compile(optimizer = 'rmsprop',
              loss = 'sparse_categorical_crossentropy',
              metrics = ['sparse_categorical_accuracy'])
    m.summary()
```
I suppose it is dividing the batch size I gave with the number of replicas available. But shouldn't that be transparent to the user? I want the code I write to work the same no matter what hw I have available (CPU, GPU or TPU).

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/cdb487d23951c4590be18bf54c77d1fb/40428.ipynb#scrollTo=sxxO4Lyy5dp-). Thanks!

**Comment by ymodak** [comment]
@bjourne In this case you may want to limit the number of replicas to 1.
```python
#strategy = TPUStrategy(resolver)
topology = tf.tpu.experimental.initialize_tpu_system(resolver)
device_assignment = tf.tpu.experimental.DeviceAssignment.build(topology, num_replicas=1)
strategy = tf.distribute.experimental.TPUStrategy(resolver, device_assignment=device_assignment)
```
By default TF maps all available cores for faster performance.

**Comment by bjourne** [comment]
I understand how it works and of course I want to use all replicas. But it is very surprising to me that the strategy would modify the model definition automatically. That is very unexpected.

**Comment by nikitamaia** [comment]
Hi @bjourne, as @ymodak mentioned, when using TPUStrategy with keras, the input dataset is assumed to be batched by the global batch size, and each batch will get divided across all the replicas. This is also the case for MirroredStrategy, and you can find [this noted in the docs here](https://www.tensorflow.org/guide/distributed_training#using_tfdistributestrategy_with_tfkerasmodelfit).

One way to think about this is to first consider the case of training on a single device, where one step is a pass through a batch, which is 32 examples in your case. In your case with TPUStrategy (or MirroredStrategy with GPUs), each replica receives 4 examples on a single step. Variables are mirrored across your devices, and the updates are calculated by aggregating the gradients calculated on each device (ie across all 32 examples). So even in the distributed case, for a single step, updates are calculated from 32 examples, as is also the case of training on a single device. Note that each epoch will then train faster as you add more devices, and typically, you would want to increase your batch size as you add more accelerators so you can make effective use of the extra computing power.

**Comment by nikitamaia** [comment]
Closing this issue now since it is not a bug.

**Comment by bjourne** [comment]
I guess it is not. You could make it clearer though because the behavior is surprising.

**Comment by nikitamaia** [comment]
Thanks for the feedback. I agree with you that it would be useful to make it clearer how batches are being divided across the replicas. It is explained in the guide, but it can still be a bit confusing.
